vault/acl.go

```diff
@@ -148,32 +221,100 @@ func (a *ACL) AllowOperation(op logical.Operation, path string) (allowed bool, s
 	if !ok {
 		return false, false
 	} else {
-		capabilities = raw.(uint32)
+		permissions = raw.(*Permissions)
+		capabilities = permissions.CapabilitiesBitmap
 	}
 
 CHECK:
 	// Check if the minimum permissions are met
 	// If "deny" has been explicitly set, only deny will be in the map, so we
 	// only need to check for the existence of other values
 	sudo = capabilities&SudoCapabilityInt > 0
+	operationAllowed := false
 	switch op {
 	case logical.ReadOperation:
-		allowed = capabilities&ReadCapabilityInt > 0
+		operationAllowed = capabilities&ReadCapabilityInt > 0
 	case logical.ListOperation:
-		allowed = capabilities&ListCapabilityInt > 0
+		operationAllowed = capabilities&ListCapabilityInt > 0
 	case logical.UpdateOperation:
-		allowed = capabilities&UpdateCapabilityInt > 0
+		operationAllowed = capabilities&UpdateCapabilityInt > 0
 	case logical.DeleteOperation:
-		allowed = capabilities&DeleteCapabilityInt > 0
+		operationAllowed = capabilities&DeleteCapabilityInt > 0
 	case logical.CreateOperation:
-		allowed = capabilities&CreateCapabilityInt > 0
+		operationAllowed = capabilities&CreateCapabilityInt > 0
 
-	// These three re-use UpdateCapabilityInt since that's the most appropriate capability/operation mapping
+	// These three re-use UpdateCapabilityInt since that's the most appropriate
+	// capability/operation mapping
 	case logical.RevokeOperation, logical.RenewOperation, logical.RollbackOperation:
-		allowed = capabilities&UpdateCapabilityInt > 0
+		operationAllowed = capabilities&UpdateCapabilityInt > 0
 
 	default:
 		return false, false
 	}
-	return
+
+	if !operationAllowed {
+		return false, sudo
+	}
+
+	// Only check parameter permissions for operations that can modify
+	// parameters.
+	if op == logical.UpdateOperation || op == logical.DeleteOperation || op == logical.CreateOperation {
+		// Check if all parameters have been denied
+		if _, ok := permissions.DeniedParameters["*"]; ok {
+			return false, sudo
+		}
+
+		_, allowedAll := permissions.AllowedParameters["*"]
+		if len(permissions.DeniedParameters) == 0 && allowedAll {
+			return true, sudo
+		}
+
+		for parameter, value := range req.Data {
+			// Check if parameter has been explictly denied
+			if valueSlice, ok := permissions.DeniedParameters[parameter]; ok {
+				// If the value exists in denied values slice, deny
+				if valueInParameterList(value, valueSlice) {
+					return false, sudo
+				}
+				// If the value doesn't exist in the denied values slice,
+				// continue
+				continue
```

Comment:
My original comment about `continue` was below in the allowed case; since it's the last check it's actually unnecessary there, but it's actually harmful here: if it's in the denied parameters list, if the value they set isn't denied, it could still (e.g. from merging different policies together) be in the allowed list with a specific set of allowed values.

If we trued up between the allowed/denied lists earlier on we could avoid this, but doing that might be both more complicated and slower. (Or maybe not, haven't thought about it in depth.)